### PR TITLE
Implement speech decoder v2

### PR DESCRIPTION
**Pull request by Edresson**

### CODE REVIEW

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py [hunk 1]
```diff
@@ -0,0 +1,1936 @@
...
+
+        # instanciate T5-TTS decoder to full compatibility and potentialy load pretrained model
+        self.t5_decoder = t5tts_transformer.Transformer(**self.speech_decoder_parms)
+
+        # projection to predict audio codes
+        self.final_proj = nn.Linear(self.speech_decoder_parms["d_model"], num_audio_codebooks * num_audio_tokens_per_codebook)
+
+        # create embeddings for encode input tokens
+        if self.cond_on_prev_audio_tokens:
+            self.audio_embedding = nn.Embedding(num_audio_tokens_per_codebook * self.num_audio_codebooks, self.speech_decoder_parms["d_model"] * self.num_audio_codebooks)
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py]: this line is not needed right?

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py [hunk 2]
```diff
@@ -0,0 +1,1936 @@
...
+            audio_embeddings = []
+            for _ in range(self.num_audio_codebooks):
+                audio_embeddings.append(nn.Embedding(num_audio_tokens_per_codebook, self.speech_decoder_parms["d_model"]))
+            self.audio_embeddings = nn.ModuleList(audio_embeddings)
+
+    def forward(self, hidden_states, speech_mask, input_audio_tokens=None):
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py]: what do you think about adding a detach() to hidden_states? both from the modeling side and from debugging this bprop speed issue

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py [hunk 3]
```diff
@@ -0,0 +1,1936 @@
...
+            embedding = self.audio_embeddings[c](audio_tokens[:, :, c])
+            if audio_embedding is None:
+                audio_embedding = embedding
+            else:
+                audio_embedding = audio_embedding + embedding
+        audio_embedding = audio_embedding / audio_tokens.size(1)
+        return audio_embedding
+
+
+class S2sMCoreGPTModelSpeechDecoder(MCoreGPTModel):
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py:314]: a note for future: after we find decoder v2 is better than v1,  i would still suggest we decouple speech decoder with MCoreGPTModel to avoid the unnecessary complexity

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py [hunk 4]
```diff
@@ -0,0 +1,1936 @@
...
+            if torch.rand(1).item() < self.cfg_unconditional_prob:
+                # make the whole batch zeros to the unconditional model
+                speech_decoder_input = torch.zeros_like(speech_decoder_input)
+                speech_mask = torch.ones_like(speech_mask)
+
+        if self.cond_on_prev_audio_tokens:
+            audio_tokens_embedded = self.embed_audio_tokens(input_audio_tokens.transpose(0, 1).contiguous()) # (B, T', E)
+            speech_decoder_input = speech_decoder_input + audio_tokens_embedded
+
+        decoder_out = self.t5_decoder(x=speech_decoder_input, x_mask=speech_mask)
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py]: iiuc t5 decoder taks BTE as input . if that's the case, you should make x= speech_decoder_input.transpose as speech_decoder_input is TBF as the mtron format

Comment (Edresson) [reply nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py]: Yes, but if  I use transpose, it will not works. The model will not converge. It is why I created the SpeechDecoderInverted class. Why it does not converge if Speech decoder works at [B, T, F] I don't know. I suspect that it is related to megatron.

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py (hunk 3 above)
Comment (Edresson) [reply nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py:296]: Yes we could try to check if it solves the issue.

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py (hunk 2 above)
Comment (Edresson) [reply nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py]: Yes we could try to check if it solves the issue.

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py (hunk 1 above)
Comment (Edresson) [reply nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py]: Yes it is not, I was using it to debug and check if the multiples audio_embeddings was the issue, but it was not the case.

File: examples/multimodal/speech_llm/conf/s2s/pt_salm_1a_21fps_debug_duplex_mistrall_spkri_speech_decoder_v2_dummy.yaml [hunk 5]
```diff
@@ -0,0 +1,479 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
```
Comment (zhehuaichen) [inline examples/multimodal/speech_llm/conf/s2s/pt_salm_1a_21fps_debug_duplex_mistrall_spkri_speech_decoder_v2_dummy.yaml:1]: why do we need this?

File: nemo/collections/multimodal/speech_llm/models/modular_s2s_models.py [hunk 6]
```diff
@@ -1152,11 +1220,32 @@ def get_num_turn(input_preds):
                         return [len(re.split('   *', pred)) for pred in input_preds]
 
                     if text_metric_name == 'bleu':  # asr-bleu, bleu
-                        metric_result = torch.Tensor([sacrebleu.corpus_bleu(text_preds, [labels]).score]).to(
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/modular_s2s_models.py:1155]: can you change to call it norm_bleu and norm_wer instead, if you include normalized text here. Or you add a flag to turn on and off this norm and make it default off. otherwise this will make the previous metric numbers cannot be reproduced

also add a "#TODO: use whisper normalizer https://pypi.org/project/whisper-normalizer/#:~:text=2%20by%20AssemblyAI-,How%20to%20use,over%20and%20pour%20me%20out%20'

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py (hunk 3 above)
Comment (zhehuaichen) [reply nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py:314]: can we add a TODO on above comment

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py [hunk 7]
```diff
@@ -0,0 +1,2052 @@
...
+                        )
+                        num_turn = 0
+                        for start, end in each_start_end_time:
+                            start = int(self.codec_sample_rate * start)
+                            end = int(self.codec_sample_rate * end)
+                            if end > start:
+                                num_turn += 1
+                                trans_new_pred_wav.append(pred_wav[start:end])
+                        num_turns.append(num_turn)
+                    asr_batch_size = min(64, len(trans_new_pred_wav))
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py:995]: add this stanza here https://github.com/zhehuaichen/NeMo/blob/speechllm-develop-gen_duplex2_clean/nemo/collections/multimodal/speech_llm/models/modular_s2s_models.py#L886-L890

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py [hunk 8]
```diff
@@ -0,0 +1,2052 @@
...
+        seconds = steps * codec_model_downsampling_factor / codec_sample_rate * decoder_reduction_factor
+        return seconds, (seconds * codec_sample_rate).int()
+
+    def get_step_from_audio_len(self, audio_len):
+        decoder_reduction_factor = self.cfg.get("decoder_reduction_factor", 1)
+        return torch.ceil(audio_len / self.codec_model_downsampling_factor / decoder_reduction_factor).int() - 1
+
+    def prepare_llm_input_duplex_from_multiturn(self, audio_batch):
+        if self.cfg.get('noise_prob', 0.0) and random.random() < self.cfg.get('noise_prob', 0.0):
+            self.add_noise_to_batch(audio_batch, os.path.join(self.cfg.noise_path, 'train'), random.randint(15, 25))
```
Comment (zhehuaichen) [inline nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py]: copy over these changes to here just for consistency https://github.com/zhehuaichen/NeMo/commit/38be2c5dcdaa86d6055cfed65fe1d7174d2edb5a

File: examples/multimodal/speech_llm/conf/s2s/pt_salm_1a_21fps_debug_duplex_mistrall_spkri_speech_decoder_v2_dummy.yaml (hunk 5 above)
Comment (Edresson) [reply examples/multimodal/speech_llm/conf/s2s/pt_salm_1a_21fps_debug_duplex_mistrall_spkri_speech_decoder_v2_dummy.yaml:1]: It is what I have been used to test the code. I reduced the number of datasets so it is faster to test.

File: nemo/collections/multimodal/speech_llm/models/modular_s2s_models.py (hunk 6 above)
Comment (Edresson) [reply nemo/collections/multimodal/speech_llm/models/modular_s2s_models.py:1155]: We already have a lot of metrics, and if we add norm_bleu, norm_bleu_asr will add extra complexity. Note that it is a bug; we should fix it anyway. I think the flag would be fine, but I would vote against the new norm_ metrics.

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py (hunk 3 above)
Comment (Edresson) [reply nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py:314]: Done

File: nemo/collections/multimodal/speech_llm/models/modular_s2s_models.py (hunk 6 above)
Comment (Edresson) [reply nemo/collections/multimodal/speech_llm/models/modular_s2s_models.py:1155]: I have added the flag norm_val_metrics.

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py (hunk 7 above)
Comment (Edresson) [reply nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py:995]: Done

File: nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py (hunk 8 above)
Comment (Edresson) [reply nemo/collections/multimodal/speech_llm/models/duplex_s2s_speech_decoder.py]: done, I copied the ones that were in the s2s model class.
